### PR TITLE
Dev heat pump migration to stack 3 to main

### PR DESCRIPTION
**Pull request by JieyangXu**
The code developed in this branch is for agents using derived information framework and for preprint 323. Details refer to individual README.

### CODE REVIEW

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Agents/CopCalculationAgent/index.html
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/CopCalculationAgent/index.html:1]: I created this page as an introduction of the calculation details of this agent, I thought separate this page will make the README more concise and navigatable. However as per your suggestions I will integrate these index in the README file.

File: Agents/CopCalculationAgent/setup.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+setup(
+    name='copcalculationa',
+    version='0.0.1',
+    author='Jieyang Xu',
+    author_email='mh807@cam.ac.uk',
+    license='MIT',
+    python_requires='>=3.8',
+    description="The `copcalculation` agent calculates the coefficient of performance for a particular month in a Lower Super Output Area in the UK \
+                and populates the result to knowledge graph as part of The World Avatar project.",
+    url="https://github.com/cambridge-cares/TheWorldAvatar/tree/dev-heat-pump-migration-to-stack-3/Agents/CopCalculationAgent",
```
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/CopCalculationAgent/setup.py:12]: Yes it was designed to switch the 'dev-heat-pump-migration-to-stack-3' into 'main' once it was awaited to be merged, now it's been done

File: Agents/CopCalculationAgent/tests/test_example_agent.py
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/CopCalculationAgent/tests/test_example_agent.py:1]: my bad, sorry. it has been removed

File: Agents/HistoricalAQMeshAgent/README.md
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/HistoricalAQMeshAgent/README.md:1]: my bad again, I made this change accidently when using global search and switch in VScode -- lesson learned!

File: Agents/LSOAInputAgent/agent/flaskapp/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+*/__pycache__
```
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/LSOAInputAgent/agent/flaskapp/.gitignore:1]: May I confirm that shall I change it into '__pycache__/' ?

File: Agents/LSOAInputAgent/deploy_agent_on_stack.html
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/LSOAInputAgent/deploy_agent_on_stack.html:1]: Noted, will do later

File: Agents/LSOAInputAgent/requirements.txt
[the remark this answers is not in the document]
Comment (JieyangXu) [reply Agents/LSOAInputAgent/requirements.txt:1]: Thank you for letting me know this update, I will remove all requirements.txt and commands in docker file . May I confirm that is all I need to do?

File: Agents/InequalityIndexCalculationAgent/README.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
+## Description
```
Comment (jwja2) [inline Agents/InequalityIndexCalculationAgent/README.md:1]: I think something that is missing at the start of each of the readme files is a statement to give some broader context. For example, that the agent was created alongside agents X, Y and Z to support the analysis of the impact of heat pumps in the UK (you could potentially cite your and Tom's preprints).

File: Agents/InequalityIndexCalculationAgent/python_scripts/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
+# Description
```
Comment (jwja2) [inline Agents/InequalityIndexCalculationAgent/python_scripts/README.md:1]: I think something that is missing at the start of each of the readme files is a statement to give some broader context. For example, that the agent was created alongside agents X, Y and Z to support the analysis of the impact of heat pumps in the UK (you could potentially cite your and Tom's preprints).

File: Agents/LSOAInputAgent/README.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
+# Description
```
Comment (jwja2) [inline Agents/LSOAInputAgent/README.md:1]: I think something that is missing at the start of each of the readme files is a statement to give some broader context. For example, that the agent was created alongside agents X, Y and Z to support the analysis of the impact of heat pumps in the UK (you could potentially cite your and Tom's preprints).

File: Agents/ResultedConsumptionCalculationAgent/README.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
+## Description
```
Comment (jwja2) [inline Agents/ResultedConsumptionCalculationAgent/README.md:1]: I think something that is missing at the start of each of the readme files is a statement to give some broader context. For example, that the agent was created alongside agents X, Y and Z to support the analysis of the impact of heat pumps in the UK (you could potentially cite your and Tom's preprints).

File: Agents/UtilityCostCalculationAgent/README.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
+## Description
```
Comment (jwja2) [inline Agents/UtilityCostCalculationAgent/README.md:1]: I think something that is missing at the start of each of the readme files is a statement to give some broader context. For example, that the agent was created alongside agents X, Y and Z to support the analysis of the impact of heat pumps in the UK (you could potentially cite your and Tom's preprints).

File: Agents/LSOAInputAgent/requirements.txt
[the remark this answers is not in the document]
Comment (sm453) [reply Agents/LSOAInputAgent/requirements.txt:1]: Removing the `requirements.txt` file and removing the command from the `Dockerfile` that installs these requirements is the first step. The second step is to check if this breaks any of your code. If it does, you will probably need to update your `twa`/`py4jps` dependencies.

File: Agents/LSOAInputAgent/agent/flaskapp/.gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (sm453) [reply Agents/LSOAInputAgent/agent/flaskapp/.gitignore:1]: `__pycache__/` in your root-level `.gitignore` file should do the job.

File: Agents/CopCalculationAgent/README.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+This agent should be implemented with other agents using Derived Information Framework. This agent works with [LSOAInputAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/LSOAInputAgent), [ResultedConsumptionCalculationAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/ResultedConsumptionCalculationAgent), [UtilityCostCalculationAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/UtilityCostCalculationAgent) and [InequalityIndexCalculationAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/InequalityIndexCalculationAgent),to support the analysis of the impact of heat pumps in the UK (Details refer to   **[Preprint 323](https://como.ceb.cam.ac.uk/preprints/323/)** with **[Preprint 281](https://como.ceb.cam.ac.uk/preprints/281/)**). You may refer to the following graph to find what is the role of this agent in these calculations:
```
Comment (jwja2) [inline Agents/CopCalculationAgent/README.md:4]: Replace "... (Details refer to..." with "... (for details refer to...". 

Same comment for other readmes if they use the same text.

File: Agents/CopCalculationAgent/README.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+![Agent framework](https://i.imgur.com/vSBvBoJ.jpeg)
+
+### Basic
+The underlying equation of the calculations in this agent is:
+
+$$
+COP = \frac{hp\_efficiency \times T_H}{T_H - T_C}
```
Comment (jwja2) [inline Agents/CopCalculationAgent/README.md:12]: There is a problem with the formatting of the subscript efficiency in this equation. Is it possible to improve it?

File: Agents/InequalityIndexCalculationAgent/python_scripts/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+The code and data collected in this folder is dedicated to reproduce the figures in **[Preprint 323](https://como.ceb.cam.ac.uk/preprints/323/): Impact of heat pumps and future energy prices on regional inequalities.**
+
+## Data
+Please download the data (should be two folder, called `Data` and `GB_shapefile` respectively) from the project folder in the Dropbox (`CoMo shared/jx309/c4e-jx309-heatpumps/Codes_and_Data`, or contact Jethro), and replace it with the `./Data` and `./GB_shapefile`.
```
Comment (jwja2) [inline Agents/InequalityIndexCalculationAgent/python_scripts/README.md:6]: Replace "(should be two folder, ..." with "(there should be two folders called..."

Comment (jwja2) [reply Agents/InequalityIndexCalculationAgent/python_scripts/README.md:6]: Replace "from the project folder in the Dropbox (`CoMo shared/jx309/c4e-jx309-heatpumps/Codes_and_Data`, or contact Jethro), and replace it with the `./Data` and `./GB_shapefile`." with "the archive for preprint 323 (available on request from the corresponding author) and use it to replace the contents of the `./Data` and `./GB_shapefile fiolders."

File: Agents/CopCalculationAgent/README.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+This agent should be implemented with other agents using Derived Information Framework. This agent works with [LSOAInputAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/LSOAInputAgent), [ResultedConsumptionCalculationAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/ResultedConsumptionCalculationAgent), [UtilityCostCalculationAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/UtilityCostCalculationAgent) and [InequalityIndexCalculationAgent](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/InequalityIndexCalculationAgent),to support the analysis of the impact of heat pumps in the UK (for details refer to   **[Preprint 323](https://como.ceb.cam.ac.uk/preprints/323/)** with **[Preprint 281](https://como.ceb.cam.ac.uk/preprints/281/)**). You may refer to the following graph to find what is the role of this agent in these calculations:
```
Comment (jwja2) [inline Agents/CopCalculationAgent/README.md:4]: Missing space in ",to".

File: Agents/InequalityIndexCalculationAgent/python_scripts/README.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
+# Description
+
+The code and data collected in this folder is dedicated to reproduce the figures in **[Preprint 323](https://como.ceb.cam.ac.uk/preprints/323/): Impact of heat pumps and future energy prices on regional inequalities.**
+
+## Data
+Please download the data (there should be two folders called `Data` and `GB_shapefile` respectively) from the project folder in the the archive for preprint 323 (available on request from the corresponding author) and use it to replace the contents of the `./Data` and `./GB_shapefile` fiolders.
```
Comment (jwja2) [inline Agents/InequalityIndexCalculationAgent/python_scripts/README.md:6]: Repeated word "the the" and typo "fiolders".

File: Agents/InequalityIndexCalculationAgent/python_scripts/README.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+│   ├── dataoutput.py # Calculation functions and core figure generation functions
+│   └── Figure_generation.py # KEY script, you will need to run this script to reproduce the figure. 
+├── dataretrieval/
+│   ├── __init__.py
+│   └── dataretrieval.py # Functions to retrieve the data.
+└── __init__.py
+Data/ # Data used in this project
+│   ├── pickles/
+│   │     └── geometry # Geometry shapes of LSOA boundaries
+│   ├── properties_csv/ # Data for each lsoa or for the UK, separated by year
```
Comment (jwja2) [inline Agents/InequalityIndexCalculationAgent/python_scripts/README.md:26]: What does "for each lsoa or for the UK" mean - could you be more precise?

For example, doe the boundaries have to be LSOAs or could they be any other type of region? Could the LSOAs (or other geometries) be for a part of the UK or the whole of the UK?

File: Agents/LSOAInputAgent/README.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+The `LSOAInput agent` is dedicated to process data around the UK Lower-layer Super Output Area (LSOA), and wrap the data as triples based on ontologies, and upload the triples into the knowledge graph.
+
+Currently this agent is focusing on the data of electricity consumption, gas consumption, fuel poverty, climate (temperature) and geometric shape to perform the use case of analysing the deployment of heat pump. The data instatiated in the knowledge graph follows[Ontoclimate](http://www.theworldavatar.com/ontology/ontogasgrid/ontoclimate.owl), [Ontofuelpoverty](http://www.theworldavatar.com/ontology/ontofuelpoverty/ontofuelpoverty.owl) and [Ontogasgrid](https://github.com/cambridge-cares/TheWorldAvatar/tree/main/GasGrid) ontologies in the [TheWorldAvatar](https://github.com/cambridge-cares/TheWorldAvatar).
```
Comment (jwja2) [inline Agents/LSOAInputAgent/README.md:5]: Missing space "follows["

Comment (jwja2) [inline Agents/LSOAInputAgent/README.md:5]: What do you mean, "Currently this agent is focusing on..."?

Can you please provide a better description of what the agent does?

File: Agents/ResultedConsumptionCalculationAgent/README.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+Resulted Gas consumption = Gas consumption - Change of gas consumption
+$$
+
+$$
+Resulted Electricity consumption = Electricity consumption + Change of electricity consumption
+$$
+
+**Arguments**:
+
+<u>Uptake</u>: a index which quantify the extent of the deployment of heat pump by measureing the gas being replaced by using heat pump comparing to conventional fossil fule heating system.
```
Comment (jwja2) [inline Agents/ResultedConsumptionCalculationAgent/README.md:35]: Could you put the arguments on separate lines? The current text is very difficult to read. Thank you.

File: Agents/UtilityCostCalculationAgent/README.md [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+Electricity Cost = Resulted Electricity Consumption × Electricity unit cost
+$$
+
+$$
+Gas Cost = Resulted Gas Consumption × Gas unit cost
+$$
+
+**Arguments**:
+
+<u>Resulted Electricity Consumption</u>: refer to resulted energy consumption calculation agents for details.
```
Comment (jwja2) [inline Agents/UtilityCostCalculationAgent/README.md:20]: Again, difficult to read. Could you improve the appearance?

File: Agents/InequalityIndexCalculationAgent/python_scripts/README.md (hunk 13 above)
Comment (JieyangXu) [reply Agents/InequalityIndexCalculationAgent/python_scripts/README.md:26]: I think it is  a good point here -- some comments may seem precise for me but confusing for reviewer, and it should be for the reviewer to judge -- let me know if there is other place that seems confusing for you,

Comment (jwja2) [reply Agents/InequalityIndexCalculationAgent/python_scripts/README.md]: I am now happy with this.

**Review by sm453 — APPROVED** [review]

**Pull request merged**